server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStore.java

```diff
@@ -40,4 +40,22 @@
     BucketName getDefaultBucketName();
 
     Mono<Void> deleteBucket(BucketName bucketName);
+
+    BucketName.Prefix bucketPrefix();
+
+    default Mono<BlobId> save(BucketName.Value bucketValue, byte[] data) {
```

Comment:
What is the difference with what is above?

Reply:
the problem here is when the vault use BlobStore, it doesn't care about the bucket prefix, but it cannot use the default BucketName, I need to either pass the value of bucket to BlobStore or find an elegant way to compute bucket name need while don't need to know bucket prefix.

I have an idea, we don't need these method, and just calculate the bucket name we need from default BucketName instance because it already carries Bucket Prefix. I will refactor this

Reply:
If i do that, and there is a case an use configures both namespace & bucketPrefix then the default bucket name doesn't carry prefix, then I'm dead.

Reply:
any suggestion?

Reply:
Yes of course! Maybe we did not carry the configuration values far enough.

It is the blobStore responsibility to resolve the defaultBucket, it is its responsibility to append the prefix as well. With both access to the prefix and to the namespace I can easily write something like this without additional API modification, and without a composite type:

```
private final Optional<String> prefix;
private final Optional<BucketName> namespace;

private BucketName (?) resolveToBackendBucketName(BucketName bucketName) {
    if (isNamespace(bucketName)) {
        return bucketName; // return the user configured value as it should be fixed for retro-compatibiblity
    }
    return prefix.map(bucketName::applyPrefix).orElse(bucketName);
}

private boolean isNamespace(BucketName bucketName) {
    return namespace.map(buketName::equals).orElse(false);
}
```

One of my criteria of acceptance here is no API level changes: it should be all contained in `blob-objectstorage` module (with minimal changes in `blob-objectstorage-guice` as well).